moveit_core/online_signal_smoothing/src/acceleration_filter.cpp

```diff
@@ -0,0 +1,299 @@
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+/* Author: Paul Gesel
+   Description: applies smoothing by limiting the acceleration between consecutive commands
+ */
+
+#include <moveit/online_signal_smoothing/acceleration_filter.h>
+#include <rclcpp/logging.hpp>
+#include <Eigen/Sparse>
+
+// Disable -Wold-style-cast because all _THROTTLE macros trigger this
+#pragma GCC diagnostic ignored "-Wold-style-cast"
+
+namespace online_signal_smoothing
+{
+// The threshold below which any velocity or position difference is considered zero.
+constexpr double COMMAND_DIFFERENCE_THRESHOLD = 1E-4;
+constexpr double ALPHA_UPPER_BOUND = 1.0;
+constexpr double ALPHA_LOWER_BOUND = 0.0;
+
+struct CSCWrapper
+{
+  std::vector<c_int> i;
+  std::vector<c_int> j;
+  std::vector<double> x;
+  csc sparse_matrix;
+
+  CSCWrapper(Eigen::SparseMatrix<double>& M)
+  {
+    M.makeCompressed();
+
+    sparse_matrix.n = M.cols();
+    sparse_matrix.m = M.rows();
+    i.assign(M.innerSize(), 0);
+    sparse_matrix.i = i.data();
+    j.assign(M.outerSize() + 1, 0);
+    sparse_matrix.p = j.data();
+    sparse_matrix.nzmax = M.nonZeros();
+    sparse_matrix.nz = -1;
+    x.assign(M.nonZeros(), 0.0);
+    sparse_matrix.x = x.data();
+
+    update(M);
+  }
+
+  void update(Eigen::SparseMatrix<double>& M)
+  {
+    for (size_t ind = 0; ind < i.size(); ++ind)
+    {
+      i[ind] = M.innerIndexPtr()[ind];
+    }
+
+    for (size_t ind = 0; ind < j.size(); ++ind)
+    {
+      j[ind] = M.outerIndexPtr()[ind];
+    }
+    for (size_t ind = 0; ind < x.size(); ++ind)
+    {
+      x[ind] = M.data().at(ind);
+    }
+  }
+};
+
+MOVEIT_STRUCT_FORWARD(OSQPDataWrapper);
+
+struct OSQPDataWrapper
+{
+  OSQPDataWrapper(Eigen::SparseMatrix<double>& P_sparse, Eigen::SparseMatrix<double>& A_sparse)
+    : P{ P_sparse }, A{ A_sparse }
+  {
+    data.n = P_sparse.rows();
+    data.m = A_sparse.rows();
+    data.P = &P.sparse_matrix;
+    q = Eigen::VectorXd::Zero(P_sparse.rows());
+    data.q = q.data();
+    data.A = &A.sparse_matrix;
+    l = Eigen::VectorXd::Zero(A_sparse.rows());
+    data.l = l.data();
+    u = Eigen::VectorXd::Zero(A_sparse.rows());
+    data.u = u.data();
+  }
+
+  void update_A(OSQPWorkspace* work, Eigen::SparseMatrix<double>& A_sparse)
+  {
+    A_sparse.makeCompressed();
+    A.update(A_sparse);
+    osqp_update_A(work, A.x.data(), OSQP_NULL, A.x.size());
+  }
+
+  CSCWrapper P;
+  CSCWrapper A;
+  Eigen::VectorXd q;
+  Eigen::VectorXd l;
+  Eigen::VectorXd u;
+  OSQPData data{};
+};
+bool AccelerationLimitedPlugin::initialize(rclcpp::Node::SharedPtr node, moveit::core::RobotModelConstPtr robot_model,
+                                           size_t num_joints)
+{
+  // copy inputs into member variables
+  node_ = node;
+  num_joints_ = num_joints;
```

Comment:
Does the number of joints differ from the information inside the robot_model? If no, maybe this member can be removed completely

Reply:
I agree that it should be extracted from the robot model. However, I think changing the plugin interface will cause a lot of issues for people. Do you think it should be done in this or a separate PR?

Reply:
Yes